spec/lib/rex/proto/kerberos/pac/krb5_pac_spec.rb

```diff
@@ -300,3 +300,86 @@
     end
   end
 end
+
+RSpec.describe Rex::Proto::Kerberos::Pac::Krb5UpnDnsInfo do
+  let(:upn) { 'test@windomain.local' }
+  let(:dns_domain_name) { 'WINDOMAIN.LOCAL' }
+  let(:sam_name) { 'test' }
+  let(:sid) { 'S-1-5-32-544' }
+
+  let(:sample) do
+    "\x28\x00\x10\x00\x1e\x00\x38\x00\x01\x00\x00\x00\x00\x00\x00\x00" \
+    "\x74\x00\x65\x00\x73\x00\x74\x00\x40\x00\x77\x00\x69\x00\x6e\x00" \
+    "\x64\x00\x6f\x00\x6d\x00\x61\x00\x69\x00\x6e\x00\x2e\x00\x6c\x00" \
+    "\x6f\x00\x63\x00\x61\x00\x6c\x00\x57\x00\x49\x00\x4e\x00\x44\x00" \
+    "\x4f\x00\x4d\x00\x41\x00\x49\x00\x4e\x00\x2e\x00\x4c\x00\x4f\x00" \
+    "\x43\x00\x41\x00\x4c\x00"
+  end
+
+  let(:sample_ext) do
+    "\x28\x00\x18\x00\x1e\x00\x40\x00\x02\x00\x00\x00\x08\x00\x5e\x00" \
+    "\x10\x00\x66\x00\x00\x00\x00\x00\x74\x00\x65\x00\x73\x00\x74\x00" \
+    "\x40\x00\x77\x00\x69\x00\x6e\x00\x64\x00\x6f\x00\x6d\x00\x61\x00" \
+    "\x69\x00\x6e\x00\x2e\x00\x6c\x00\x6f\x00\x63\x00\x61\x00\x6c\x00" \
+    "\x57\x00\x49\x00\x4e\x00\x44\x00\x4f\x00\x4d\x00\x41\x00\x49\x00" \
+    "\x4e\x00\x2e\x00\x4c\x00\x4f\x00\x43\x00\x41\x00\x4c\x00\x74\x00" \
+    "\x65\x00\x73\x00\x74\x00\x01\x02\x00\x00\x00\x00\x00\x05\x20\x00" \
+    "\x00\x00\x20\x02\x00\x00"
+  end
+
+  context 'with non-extended upn dns info' do
+    describe '#read' do
```

Comment:
A `#write` example would be good to add too, from a BinData object that doesn't have the offsets specified upfront :+1:

Reply:
We talked about this already but just to leave a paper trail I think this covers what you're asking for here:
https://github.com/rapid7/metasploit-framework/pull/17603/files#diff-4c294da130b0102eb5a73e30dc42d56ae8ac3bb3dc990dc7d41b230e2c956437R371-R381

Reply:
I think I'm more used to seeing this sort of pattern that uses `described_class` instead, in conjunction with initialising the object values directly in the constructor instead of mutating `subject`

https://github.com/rapid7/metasploit-framework/blob/master/spec/lib/rex/proto/kerberos/credential_cache/krb5_ccache_spec.rb#L45-L54

Reply:
ok I think I've done what you're asking for, honestly didn;t even see that you'd left the comment on `#read` totally my fault for not realising I'd done that too